modules/nixos/common/update.nix

```
@@ -0,0 +1,52 @@
+{ config, pkgs, ... }:
+{
+  # adapted from:
+  # https://github.com/Mic92/dotfiles/blob/020180880d9413e076073889f82c4751a27734e9/nixos/modules/update-prefetch.nix
+  # https://github.com/NixOS/nixpkgs/blob/3428bdf3c93a7608615dddd44dec50c3df89b4be/nixos/modules/system/boot/kexec.nix
+  # https://github.com/NixOS/nixpkgs/blob/3428bdf3c93a7608615dddd44dec50c3df89b4be/nixos/modules/tasks/auto-upgrade.nix
+  systemd.services.update-host = {
+    restartIfChanged = false;
+    unitConfig.X-StopOnRemoval = false;
+    serviceConfig.Restart = "on-failure";
+    serviceConfig.RestartSec = "1m";
+    serviceConfig.Type = "oneshot";
+    path = [
+      config.nix.package
+      config.systemd.package
+      pkgs.coreutils
+      pkgs.curl
+      pkgs.kexec-tools
+    ];
+    script = ''
+      hostname=$(</proc/sys/kernel/hostname)
+      system="$(curl -sL https://buildbot.nix-community.org/nix-outputs/nix-community/infra/nixos-$hostname)"
+
+      if [[ "$(readlink -f /nix/var/nix/profiles/system)" == "$system" ]]; then
+        return
+      fi
+
+      nix-store --realise $system
+      nix-env --profile /nix/var/nix/profiles/system --set $system
+      /nix/var/nix/profiles/system/bin/switch-to-configuration boot
+
+      booted="$(readlink /run/booted-system/{initrd,kernel,kernel-modules} && cat /run/booted-system/kernel-params)"
+      p="$(readlink -f /nix/var/nix/profiles/system)"
+      built="$(readlink $p/{initrd,kernel,kernel-modules} && cat $p/kernel-params)"
+      if [[ "$booted" != "$built" ]]; then
+        # don't use kexec if system is virtualized
```

I guess this is the reason?

```suggestion
        # don't use kexec if system is virtualized, reboots are fast enough
```